File: web-sites-hub/apple-ds-core-docs/11-trie/trie.py
"""
字典树（Trie）实现
"""

class TrieNode:
    """Trie节点"""
    
    def __init__(self):
        self.children = {}
        self.is_end = False


class Trie:
    """字典树实现"""
    
    def __init__(self):
        self.root = TrieNode()
    
    def insert(self, word):
        """插入单词"""
        node = self.root
        for char in word:
            if char not in node.children:
                node.children[char] = TrieNode()
            node = node.children[char]
        node.is_end = True
    
    def search(self, word):
        """查找完整单词"""
        node = self.root
        for char in word:
            if char not in node.children:
                return False
            node = node.children[char]
        return node.is_end
    
    def starts_with(self, prefix):
        """查找前缀"""
        node = self.root
        for char in prefix:
            if char not in node.children:
                return False
            node = node.children[char]
        return True
    
    def get_all_words_with_prefix(self, prefix):
        """获取所有以prefix开头的单词"""
        node = self.root
        for char in prefix:
            if char not in node.children:
                return []
            node = node.children[char]
        
        result = []
        self._dfs(node, prefix, result)
        return result
    
    def _dfs(self, node, path, result):
        """DFS收集所有单词"""
        if node.is_end:
            result.append(path)
        
        for char, child in node.children.items():
            self._dfs(child, path + char, result)
    
    def delete(self, word):
        """删除单词"""
        def delete_helper(node, word, index):
            if index == len(word):
                if not node.is_end:
                    return False
                node.is_end = False
                return len(node.children) == 0
            
            char = word[index]
            if char not in node.children:
                return False
            
            should_delete = delete_helper(node.children[char], word, index + 1)
            
            if should_delete:
                del node.children[char]
                return not node.is_end and len(node.children) == 0
            
            return False
        
        delete_helper(self.root, word, 0)


def demo():
    """演示Trie操作"""
    print("=== 字典树演示 ===\n")
    
    trie = Trie()
    
    # 插入单词
    words = ["cat", "car", "card", "care", "dog", "dodge"]
    print(f"插入单词: {words}\n")
    for word in words:
        trie.insert(word)
    
    # 查找
    print("查找:")
    test_words = ["cat", "can", "car", "card"]
    for word in test_words:
        found = trie.search(word)
        print(f"  {word}: {'✅找到' if found else '❌未找到'}")
    
    # 前缀查找
    print("\n前缀查找:")
    prefixes = ["ca", "do", "da"]
    for prefix in prefixes:
        found = trie.starts_with(prefix)
        print(f"  {prefix}: {'✅存在' if found else '❌不存在'}")
    
    # 自动补全
    print("\n自动补全:")
    prefix = "car"
    words = trie.get_all_words_with_prefix(prefix)
    print(f"  以'{prefix}'开头的单词: {words}")


if __name__ == '__main__':
    demo()

File: web-sites-hub/frontend-docs/apple-ds-core-docs/11-trie/trie.cpp
/**
 * 字典树（Trie）实现
 */

#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

class TrieNode {
public:
    unordered_map<char, TrieNode*> children;
    bool isEnd;
    
    TrieNode() : isEnd(false) {}
};

class Trie {
private:
    TrieNode* root;
    
    void deleteTree(TrieNode* node) {
        if (node) {
            for (auto& [ch, child] : node->children) {
                deleteTree(child);
            }
            delete node;
        }
    }
    
    void dfsCollect(TrieNode* node, const string& path, vector<string>& result) {
        if (node->isEnd) {
            result.push_back(path);
        }
        
        for (auto& [ch, child] : node->children) {
            dfsCollect(child, path + ch, result);
        }
    }

public:
    Trie() {
        root = new TrieNode();
    }
    
    ~Trie() {
        deleteTree(root);
    }
    
    // 插入单词
    void insert(const string& word) {
        TrieNode* node = root;
        for (char ch : word) {
            if (!node->children.count(ch)) {
                node->children[ch] = new TrieNode();
            }
            node = node->children[ch];
        }
        node->isEnd = true;
    }
    
    // 查找完整单词
    bool search(const string& word) {
        TrieNode* node = root;
        for (char ch : word) {
            if (!node->children.count(ch)) {
                return false;
            }
            node = node->children[ch];
        }
        return node->isEnd;
    }
    
    // 查找前缀
    bool startsWith(const string& prefix) {
        TrieNode* node = root;
        for (char ch : prefix) {
            if (!node->children.count(ch)) {
                return false;
            }
            node = node->children[ch];
        }
        return true;
    }
    
    // 获取所有以prefix开头的单词
    vector<string> getAllWordsWithPrefix(const string& prefix) {
        TrieNode* node = root;
        for (char ch : prefix) {
            if (!node->children.count(ch)) {
                return {};
            }
            node = node->children[ch];
        }
        
        vector<string> result;
        dfsCollect(node, prefix, result);
        return result;
    }
};

int main() {
    cout << "=== 字典树演示 ===" << endl << endl;
    
    Trie trie;
    
    // 插入单词
    vector<string> words = {"cat", "car", "card", "care", "dog", "dodge"};
    cout << "插入单词: ";
    for (const auto& word : words) {
        cout << word << " ";
        trie.insert(word);
    }
    cout << endl << endl;
    
    // 查找
    cout << "查找:" << endl;
    vector<string> testWords = {"cat", "can", "car", "card"};
    for (const auto& word : testWords) {
        bool found = trie.search(word);
        cout << "  " << word << ": " << (found ? "✅找到" : "❌未找到") << endl;
    }
    cout << endl;
    
    // 前缀查找
    cout << "前缀查找:" << endl;
    vector<string> prefixes = {"ca", "do", "da"};
    for (const auto& prefix : prefixes) {
        bool found = trie.startsWith(prefix);
        cout << "  " << prefix << ": " << (found ? "✅存在" : "❌不存在") << endl;
    }
    cout << endl;
    
    // 自动补全
    string prefix = "car";
    cout << "自动补全:" << endl;
    vector<string> completions = trie.getAllWordsWithPrefix(prefix);
    cout << "  以'" << prefix << "'开头的单词: ";
    for (const auto& word : completions) {
        cout << word << " ";
    }
    cout << endl;
    
    return 0;
}

